File: python/test/t_FunctionalChaos_gsobol_sparse.py
#! /usr/bin/env python

from openturns import *
from math import *

TESTPREAMBLE()

try :
    #Log.Show( Log.Flags() | Log.INFO )

      # Problem parameters
    dimension = 8

    # Create the Ishigami function
    # Reference analytical values
    meanTh = 1.0
    covTh = 1.0
    a = NumericalPoint(dimension)
    a[0] = 1.0
    a[1] = 2.0
    a[2] = 5.0
    a[3] = 10.0
    a[4] = 20.0
    a[5] = 50.0
    a[6] = 100.0
    a[7] = 500.0
    inputVariables = Description(dimension)
    outputVariables = Description(1)
    outputVariables[0] = "y"
    formula = Description(1)
    formula[0] = "1.0"
    for i in range(dimension):
        covTh = covTh * (1.0 + 1.0 / (3.0 * (1.0 + a[i])**2))
        inputVariables[i] = "xi"+str(i)
        formula[0] = formula[0] + " * ((abs(4.0 * xi" + str(i) + " - 2.0) + " + str(a[i]) + ") / (1.0 + " + str(a[i]) + "))"
    covTh = covTh - 1.0

    model = NumericalMathFunction(inputVariables, outputVariables, formula)

    # Create the input distribution
    marginalX = DistributionCollection(dimension)
    for i in range(dimension):
        marginalX[i] = Uniform(0, 1.0)
    distribution = ComposedDistribution(marginalX)

    # Create the orthogonal basis
    polynomialCollection = PolynomialFamilyCollection(dimension)
    for i in range(dimension):
        polynomialCollection[i] = OrthogonalUniVariatePolynomialFamily(LegendreFactory())
    q = 0.4
    enumerateFunction = HyperbolicAnisotropicEnumerateFunction(dimension, q)
    productBasis = OrthogonalBasis(OrthogonalProductPolynomialFactory(polynomialCollection, enumerateFunction))


    # design experiment
    samplingSize = 100

    # build basis
    basisSize = 917

    listFittingAlgorithm = list()
    listFittingAlgorithm.append(KFold())
    listFittingAlgorithm.append(CorrectedLeaveOneOut())

    for fittingAlgorithmIndex in range(len(listFittingAlgorithm)):
        fittingAlgorithm = listFittingAlgorithm[fittingAlgorithmIndex]
        adaptiveStrategy = FixedStrategy( productBasis, basisSize )
        projectionStrategy = LeastSquaresStrategy( LowDiscrepancyExperiment( SobolSequence(dimension), samplingSize), LeastSquaresMetaModelSelectionFactory ( LAR(), fittingAlgorithm) );
        algo = FunctionalChaosAlgorithm(model, distribution, adaptiveStrategy, projectionStrategy)
        RandomGenerator.SetSeed(0)
        algo.run()

        result = algo.getResult()
        print "coeffs = ", result.getCoefficients()
        print "residuals = ", result.getResiduals()
        print "relative errors = ", result.getRelativeErrors()



except :
    import sys
    print "t_LeastSquaresMetaModelSelection_std.py", sys.exc_type, sys.exc_value

File: lib/test/t_FunctionalChaos_gsobol_sparse.cxx
//                                               -*- C++ -*-
/**
 *  @file  t_LeastSquaresMetaModelSelection_std.cxx
 *  @brief The test file of class LeastSquaresMetaModelSelection for standard methods
 *
 *  Copyright (C) 2005-2013 EDF-EADS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  @author schueller
 *  @date   2010-02-04 16:44:49 +0100 (Thu, 04 Feb 2010)
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  //   Log::Show( Log::Flags() | Log::INFO );

  // Problem parameters
  UnsignedLong dimension(8);

  // Reference analytical values
  NumericalScalar covTh(1.0);
  NumericalPoint a(dimension);
  a[0] = 1.0;
  a[1] = 2.0;
  a[2] = 5.0;
  a[3] = 10.0;
  a[4] = 20.0;
  a[5] = 50.0;
  a[6] = 100.0;
  a[7] = 500.0;

  // Create the gSobol function
  Description inputVariables(dimension);
  Description outputVariables(1);
  outputVariables[0] = "y";
  Description formula(1);
  formula[0] = "1.0";
  for (UnsignedLong i = 0; i < dimension; ++i)
    {
      covTh *= 1.0 + 1.0 / (3.0 * pow(1.0 + a[i], 2.0));
      inputVariables[i] = (OSS() << "xi" << i);
      formula[0] = (OSS() << formula[0] << " * ((abs(4.0 * xi" << i << " - 2.0) + " << a[i] << ") / (1.0 + " << a[i] << "))");
    }
  --covTh;

  NumericalMathFunction model(inputVariables, outputVariables, formula);

  // Create the input distribution
  Collection<Distribution> marginalX(dimension);
  for ( UnsignedLong i = 0; i < dimension; ++ i )
    {
      marginalX[i] = Uniform( 0.0, 1.0 );
    }
  ComposedDistribution distribution(marginalX);

  // Create the orthogonal basis
  Collection<OrthogonalUniVariatePolynomialFamily> polynomialCollection(dimension);
  for ( UnsignedLong i = 0; i < dimension; ++ i )
    {
      polynomialCollection[i] = LegendreFactory();
    }
  const NumericalScalar q(0.4);
  HyperbolicAnisotropicEnumerateFunction enumerateFunction(dimension, q);
  OrthogonalProductPolynomialFactory productBasis(polynomialCollection, enumerateFunction);

  // design experiment
  UnsignedLong samplingSize( 100 );

  // build basis
  UnsignedLong basisSize( 917 );

  // run algorithm
  Collection<FittingAlgorithm> listFittingAlgorithm;
  listFittingAlgorithm.add(KFold());
  listFittingAlgorithm.add(CorrectedLeaveOneOut());

  for ( UnsignedLong fittingAlgorithmIndex = 0; fittingAlgorithmIndex < listFittingAlgorithm.getSize(); ++ fittingAlgorithmIndex )
    {
      FixedStrategy adaptiveStrategy( productBasis, basisSize );
      FittingAlgorithm fittingAlgorithm( listFittingAlgorithm[fittingAlgorithmIndex] );
      LeastSquaresStrategy projectionStrategy( LowDiscrepancyExperiment(SobolSequence(dimension), samplingSize), LeastSquaresMetaModelSelectionFactory ( LAR(), fittingAlgorithm) );
      FunctionalChaosAlgorithm algo(model, distribution, adaptiveStrategy, projectionStrategy);
      RandomGenerator::SetSeed(0);
      algo.run();

      FunctionalChaosResult result(algo.getResult());
      fullprint << "coeffs = " << result.getCoefficients() << std::endl;
      fullprint << "residuals = " << result.getResiduals() << std::endl;
      fullprint << "relative errors = " << result.getRelativeErrors() << std::endl;
    }




  return ExitCode::Success;
}
